### PR TITLE
chore: add unit tests for untested daos

### PR DESCRIPTION
**Pull request by harbassan**
## Issue

Much of the dao logic was completely untested.

## Solution

Added many unit tests for a large percentage of dao logic.

## Risk

Funcionality of some daos was slightly modified as tests revealed issues, but no changes are logic changes.

## Checklist

- [ ] Acceptance criteria met
- [ ] Wiki documentation is written and up to date
- [ ] Unit tests written and passing
- [ ] Integration tests written and passing
- [ ] Continuous integration build passing

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved not-found errors for missing groups, scenarios, and users.
  - Improved scene validation, ordering checks, visit tracking, and file-reference updates.
  - Database persistence errors are now surfaced instead of silently ignored.

- **Changes**
  - Scenario updates no longer accept duration changes.
  - Removed the endpoint for retrieving played users.

- **Documentation**
  - Clarified documentation for data and access operations.

- **Tests**
  - Expanded automated coverage for data access, edge cases, and error handling.

### CODE REVIEW

**Comment by harbassan** [comment]
@coderabbitai review

File: backend/src/db/daos/sceneDao.js [hunk 1]
```diff
@@ -89,25 +125,75 @@ const assertDirectLinkInScenario = async (scenarioId, directLinkId) => {
...
+    if (scenarioUpdateSucceeded) {
+      await Scenario.updateOne(
+        { _id: scenarioId },
+        { $pull: { scenes: dbScene._id } }
+      ).catch(() => {});
+    }
+
+    if (dbScene._id && !scenarioUpdateSucceeded) {
+      await Scene.deleteOne({ _id: dbScene._id }).catch(() => {});
+    }
```
Comment (leowla) [inline backend/src/db/daos/sceneDao.js]: it would be safer to always try to delete the scene if an error occurs since there might be partial writes

File: backend/src/db/daos/fileDao.js [hunk 2]
```diff
@@ -26,16 +27,27 @@ export async function applyReferenceDeltas(fileRefDeltas) {
...
+  try {
+    const result = await UploadedFile.bulkWrite(fileOperations, {
+      ordered: false,
+    });
+    const hasWriteErrors = Array.isArray(result?.writeErrors)
+      ? result.writeErrors.length > 0
+      : false;
+    return result?.ok === 1 && !hasWriteErrors;
+  } catch {
+    return false;
```
[the remark this answers is not in the document]
Comment (harbassan) [reply backend/src/db/daos/fileDao.js]: this is out of scope since it requires a transaction to be done well, and that will be addressed later in this codebase. this is resolved for now.

[the remark this answers is not in the document]
Comment (harbassan) [reply backend/src/db/daos/fileDao.js]: no, consider the issue resolved

File: backend/src/db/daos/sceneDao.js [hunk 3]
```diff
@@ -89,25 +125,75 @@ const assertDirectLinkInScenario = async (scenarioId, directLinkId) => {
...
+
+    if (fileReferenceUpdateSucceeded && scenarioUpdateSucceeded) {
```
Comment (leowla) [inline backend/src/db/daos/sceneDao.js]: `fileReferenceUpdateSucceeded` will always be false if an error occurs

**Comment by harbassan** [comment]
@leowla i've just removed the rollback logic entirely. it was a bad solution, and there's no point finaggling with that when the best solution is clearly to use transactions. will make a ticket for that now.

File: backend/src/db/daos/scenarioDao.js [hunk 4]
```diff
@@ -179,16 +193,25 @@ const updateScenario = async (scenarioId, updatedScenario) => {
  * @param {updatedDuration: Object} updatedDurations updated duration for a user
  * @returns updated database scene object
  */
-const updateDurations = async (scenarioId, updatedDurations) => {
-  // if we are updating name only, components will be null
+export const updateDurations = async (scenarioId, updatedDurations) => {
+  const scenario = await getScenarioOrThrow(scenarioId);
+  scenario.durations = [...(scenario.durations ?? []), updatedDurations];
```
Comment (leowla) [inline backend/src/db/daos/scenarioDao.js]: `durations` doesn't exist in the scenario schema

File: backend/src/db/daos/scenarioDao.js [hunk 5]
```diff
@@ -296,7 +315,7 @@ const deleteScenario = async (scenarioId) => {
  * @param {String} sceneId MongoDB ID of scene
  * @returns state variables for the scenario
  */
-const getStateVariables = async (scenarioId) => {
+export const getStateVariables = async (scenarioId) => {
   const scenario = await Scenario.findById(scenarioId);
   return scenario.stateVariables || [];
 };
```
Comment (leowla) [inline backend/src/db/daos/scenarioDao.js]: here and many other places seem like a good place to have your new `getScenarioOrThrow`

File: backend/src/db/daos/__tests__/groupDao.test.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      setGroupStateVariables(new mongoose.Types.ObjectId().toString(), [
+        { id: "missing", name: "hp", value: 1 },
+      ])
+    ).rejects.toMatchObject({
+      status: 404,
+      message: "group not found",
+    });
+  });
+
+  it("creates a group and removes a matching user by email", async () => {
```
Comment (leowla) [inline backend/src/db/daos/__tests__/groupDao.test.js:50]: this is not checking escaped regex e.g. `alice@example\.com`

File: backend/src/db/daos/__tests__/userDao.test.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  it("adds the scenario id to users with and without existing assignments", async () => {
+    const withAssignment = await User.create({
+      uid: "firebase-user-assign-1",
+      name: "Assigned user",
+      email: "assigned@example.com",
+      pictureURL: "https://example.com/assigned.png",
+      assigned: ["existing-scenario"],
+    });
+
+    const withoutAssignment = await User.create({
```
Comment (leowla) [inline backend/src/db/daos/__tests__/userDao.test.js]: idk mongoose but claude says it User.create will init the `assigned` field with a blank array instead of actually being empty which is not expected behaviour in this test

**Review by leowla — CHANGES_REQUESTED** [review]

**Comment by harbassan** [comment]
damn i told claude code to make no mistakes 😞

File: backend/src/db/daos/sceneDao.js [hunk 8]
```diff
@@ -89,16 +125,28 @@ const assertDirectLinkInScenario = async (scenarioId, directLinkId) => {
...
   await dbScene.save();
 
   await Scenario.updateOne(
     { _id: scenarioId },
     { $push: { scenes: dbScene._id } }
   );
 
+  await applyReferenceDeltas(fileRefDeltas);
+
   return dbScene;
```
[the remark this answers is not in the document]
Comment (harbassan) [reply backend/src/db/daos/sceneDao.js:178]: this will be done as a separate pr addressing transaction implementation across the api, and is therefore not in scope for this pr. consider this resolved

**Comment by harbassan** [comment]
@coderabbitai approve

File: backend/src/db/daos/fileDao.js [hunk 9]
```diff
@@ -26,16 +27,23 @@ export async function applyReferenceDeltas(fileRefDeltas) {
...
-  if (fileOperations.length > 0) {
-    await UploadedFile.bulkWrite(fileOperations, { ordered: false });
+  if (fileOperations.length === 0) return;
+
+  const result = await UploadedFile.bulkWrite(fileOperations, {
+    ordered: false,
+  });
+
+  if (result.matchedCount !== fileOperations.length) {
+    throw new Error("one or more file reference updates did not match");
```
Comment (leowla) [inline backend/src/db/daos/fileDao.js:37]: an error would be misleading because it does not stop or revert the writes to a scene

File: backend/src/db/daos/scenarioDao.js
Comment (leowla) [inline backend/src/db/daos/scenarioDao.js:284]: prob out of scope but also need to delete everything in `scene.background` too since its not a component

File: backend/src/db/daos/fileDao.js (hunk 9 above)
Comment (harbassan) [reply backend/src/db/daos/fileDao.js:37]: what would your suggestion be (for rn, considering we would add transactions at some point)

Comment (leowla) [reply backend/src/db/daos/fileDao.js:37]: oh yeah prob just leave it i forgot about transactions coming

File: backend/src/db/daos/scenarioDao.js
Comment (harbassan) [reply backend/src/db/daos/scenarioDao.js:284]: yeah out of scope for this one, ill note it down though for a future ticket
